test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/PreserveTypesAndAttributesTests.cs

```diff
@@ -0,0 +1,395 @@
+//---------------------------------------------------------------------
+// <copyright file="PreserveTypesAndAttributesTests.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.OData.Edm;
+using Xunit;
+
+namespace Microsoft.OData.Client.Tests.ALinq
+{
+    /// <summary>
+    /// Tests to check original resource type is used when retrieving attributes like OriginalNameAttribute
+    /// even if the IQueryable<T> object is casted to another type.
+    /// </summary>
+    public class PreserveTypesAndAttributesTests
```

Comment:
@joecarl Disabuse yourself from the notion that each test needs it's own instance of `DataServiceContext`. If that was the case and you needed to test 25 scenarios, you'd end up with 25 `DataServiceContext` instances in the `TestContext` class.
I'll reiterate what I said in your #2862 pull request. You don't need the `TestContext` class at all.

Below is a rewrite of the first 2 tests without the `TestContext` class.

The naming for the test method should give someone an idea what scenario is being tested - `OneLevelCase1` and `OneLevelCase2` do not do that...
```csharp
namespace Microsoft.OData.Client.Tests.ALinq
{
    public class PreserveTypesAndAttributesTests // You can make this class implement IDisposable if you feel very strongly about removing the SendingRequest2 delegate...
    {
        private readonly DataServiceContext ctx;
        private readonly EdmModel model;

        public PreserveTypesAndAttributesTests()
        {
            this.model = this.BuildEdmModel();
            this.ctx = new DataServiceContext(new Uri("https://mock.odata.service"));
            this.ctx.Format.UseJson(this.model);
            this.ctx.ResolveName = (type) => $"NS.{type.Name}";
            this.ctx.MergeOption = MergeOption.NoTracking;
        }

        [Fact]
        public void TestMethodNameThatGivesSomeoneABetterIdeaWhatScenarioIsBeingTested1()
        {
            IQueryable<ICustomer> query1 = this.ctx.CreateQuery<OneLevelCustomer1>("Customers")
                .Where(c => c.CorporationName != "").OrderBy(c => c.Id).Take(2);

            IQueryable<OneLevelCustomer1> query2 = this.ctx.CreateQuery<OneLevelCustomer1>("Customers")
                .Where(c => c.CorporationName != "").OrderBy(c => c.Id).Take(2);

            // TODO: Include the `@odata.context` control information property in the payload
            string resp = "[{\"CustomerID\":\"ALFoKI\",\"CompanyName\":\"Alfreds Futterkiste\",\"ContactName\":\"Maria Anders\",\"City\":\"Berlin\",\"Region\":null},{\"CustomerID\":\"ANATR\",\"CompanyName\":\"Ana Trujillo Emparedados y helados\",\"ContactName\":\"Ana Trujillo\",\"City\":\"M\\u00e9xico D.F.\",\"Region\":null}]";

            InterceptRequestAndMockResponseValue("Customers", resp);
            this.ctx.SendingRequest2 += (sender, e) =>
            {
                Assert.EndsWith("<EXPECTED_URL>", e.RequestMessage.Url.OriginalString);
            };

            var result1 = query1.ToList();
            var result2 = query2.ToList();

            Assert.Equal(result1.Count, result2.Count);
            for (var i = 0; i < result1.Count; i++)
            {
                Assert.Equal(result1[i].City, result2[i].City);
                Assert.Equal(result1[i].CorporationName, result2[i].CorporationName);
                Assert.Equal(result1[i].Name, result2[i].Name);
                Assert.Equal(result1[i].Id, result2[i].Id);
            }
        }

        [Fact]
        public void TestMethodNameThatGivesSomeoneABetterIdeaWhatScenarioIsBeingTested2()
        {
            IQueryable<BaseCustomer> query1 = ctx.CreateQuery<OneLevelCustomer2>("Customers");
                .Where(c => c.Id != "").OrderBy(c => c.Id).Select(c => new { c.Id }).Take(2);

            IQueryable<OneLevelCustomer2> query2 = ctx.CreateQuery<OneLevelCustomer2>("Customers")
                .Where(c => c.Id != "").OrderBy(c => c.Id).Select(c => new { c.Id }).Take(2);

            // TODO: Include the `@odata.context` control information property in the payload
            string resp = "[{\"CustomerID\":\"ALFKI\"},{\"CustomerID\":\"ANATR\"}]";

            this.ctx.InterceptRequestAndMockResponseValue("Customers(CustomerID)", resp);
            this.ctx.SendingRequest2 += (sender, e) =>
            {
                Assert.EndsWith("<EXPECTED_URL>", e.RequestMessage.Url.OriginalString);
            };

            var result1 = query1s.ToList();
            var result2 = query2s.ToList();

            Assert.Equal(result1.Count, result2.Count);
            for (var i = 0; i < result1.Count; i++)
            {
                Assert.Equal(result1[i].Id, result2[i].Id);
            }
        }

        // You can put the private helper methods in a #region to make the section optionally collapsible

        // You don't need to pass a DataServiceContext instance since `ctx` object is accessible to this method
        private void InterceptRequestAndMockResponse(string mockResponse)
        {
            // ctx.ResolveName = (type) => $"NS.{type.Name}"; <= DOES NOT BELONG HERE
            // THE REST OF THE CODE ...
        }

       // You don't need the other InterceptRequestAndMockResponse overload. Include the `@odata.context` control information property in the payload (from the test)

        private static EdmModel BuildEdmModel()
        {
            // CODE ..
        }
    }
}
```

Reply:
I refactored the whole code following every single indication, even those which from my subjective point of view provide questionable improvement.
Also, fortunately in this case there's no need to add a listener to `SendingRequest2` since we can assert equality between each query's request uri, as I was already doing.